### ISSUE TITLE
How to test more servers

### ISSUE THREAD

**Issue by iris-garcia** [joaotavora/eglot#595]
I would like to add more language servers to the list of Eglot, but I'm not sure how to debug them.

I started with groovy-language-server:
```
(add-to-list 'eglot-server-programs '(groovy-mode . ("java" "-jar" "/path/to/groovy-language-server/groovy-language-server-all.jar")))
```

And I can't see for example completions, is there a guideline on how to debug this?

Thank you.

**Comment by joaotavora** [comment]
Does the `*EGLOT (<project>/<major mode>) events*` buffer hold any clues. That's usually how I do it.  Also, you must make sure that the line `java -jar blablabla` runs correctly in the context of the file you are using.  For a quick test, you can use the keybinding `M-!`, or `M-x shell-command`.  If your get some error output, it will show up there.

Eglot communicates via stdin/stdout and that `M-x shell-command` trick has stdin closed on start. So, instead of `M-!` you can also just `M-x shell` and then type the command in the shell you're using.  If that program seems to hang waiting for input, it's a good thing.  In the real-world usage it will be Eglot feeding it this input.

So, to summarize:

1. `*EGLOT (<project>/<major-mode>) events*` buffer
2. `M-!` (`M-x shell-command`)
3. `M-x shell`
4. tracing functions, debugging fuctions.  In general, programming

**Comment by iris-garcia** [comment]
First of all, thanks for the detailed answer.
I think I have the language server running properly, since I can see this in the events:
```
[client-request] (id:1) Fri Jan 15 12:37:08 2021:
(:jsonrpc "2.0" :id 1 :method "initialize" :params
          (:processId 4125852 :rootPath "/tmp/groovy/" :rootUri "file:///tmp/groovy/" :initializationOptions #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
                                                                                                                           ())
                      :capabilities
                      (:workspace
                       (:applyEdit t :executeCommand
                                   (:dynamicRegistration :json-false)
                                   :workspaceEdit
                                   (:documentChanges :json-false)
                                   :didChangeWatchedFiles
                                   (:dynamicRegistration t)
                                   :symbol
                                   (:dynamicRegistration :json-false)
                                   :configuration t)
                       :textDocument
                       (:synchronization
                        (:dynamicRegistration :json-false :willSave t :willSaveWaitUntil t :didSave t)
                        :completion
                        (:dynamicRegistration :json-false :completionItem
                                              (:snippetSupport t)
                                              :contextSupport t)
                        :hover
                        (:dynamicRegistration :json-false :contentFormat
                                              ["markdown" "plaintext"])
                        :signatureHelp
                        (:dynamicRegistration :json-false :signatureInformation
                                              (:parameterInformation
                                               (:labelOffsetSupport t)))
                        :references
                        (:dynamicRegistration :json-false)
                        :definition
                        (:dynamicRegistration :json-false)
                        :declaration
                        (:dynamicRegistration :json-false)
                        :implementation
                        (:dynamicRegistration :json-false)
                        :typeDefinition
                        (:dynamicRegistration :json-false)
                        :documentSymbol
                        (:dynamicRegistration :json-false :hierarchicalDocumentSymbolSupport t :symbolKind
                                              (:valueSet
                                               [1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26]))
                        :documentHighlight
                        (:dynamicRegistration :json-false)
                        :codeAction
                        (:dynamicRegistration :json-false :codeActionLiteralSupport
                                              (:codeActionKind
                                               (:valueSet
                                                ["quickfix" "refactor" "refactor.extract" "refactor.inline" "refactor.rewrite" "source" "source.organizeImports"]))
                                              :isPreferredSupport t)
                        :formatting
                        (:dynamicRegistration :json-false)
                        :rangeFormatting
                        (:dynamicRegistration :json-false)
                        :rename
                        (:dynamicRegistration :json-false)
                        :publishDiagnostics
                        (:relatedInformation :json-false))
                       :experimental #s(hash-table size 65 test eql rehash-size 1.5 rehash-threshold 0.8125 data
                                                   ()))))
[server-reply] (id:1) Fri Jan 15 12:37:08 2021:
(:jsonrpc "2.0" :id 1 :result
          (:capabilities
           (:textDocumentSync 1 :hoverProvider t :completionProvider
                              (:resolveProvider :json-false :triggerCharacters
                                                ["."])
                              :signatureHelpProvider
                              (:triggerCharacters
                               ["(" ","])
                              :definitionProvider t :typeDefinitionProvider t :referencesProvider t :documentSymbolProvider t :workspaceSymbolProvider t :renameProvider t)))
```

What I don't know is how to interpret these capabilities, because with this output I would expect to have completion of variables/functions names for example.

**Comment by iris-garcia** [comment]
Nevermind, it was an issue with my company setup, it does return completions, thank you!
